code/src/main/kotlin/builtins/HclBuiltinFunctions.kt

```diff
@@ -6,25 +6,123 @@ import parser.AstNode.Type
+
+                    buildOperatorTxtTxtToBool("=="),
+                    buildOperatorTxtTxtToBool("!="),
+
+                    buildOperatorBoolBoolToBool("&&"),
+                    buildOperatorBoolBoolToBool("||"),
+
+                    buildPrefixOperator<Type.Bool, Type.Bool>("negated", "!")
+            ) +
+            // Control structures
+            listOf(
+                    buildThenFunction(),
+                    buildWhileFunction()
+            ) +
+            // Standard functions
+            listOf(
+                    buildToStringFunction<Type.Number>(),
+                    buildToStringFunction<Type.Text>(), //Redundant, but no reason for compiler to throw an error
+                    buildToStringFunction<Type.Bool>(),
+                    buildGetListSizeFunction()
+            )
 }
 
-private fun buildOperator(functionName: String, operator: String = functionName) = buildFunction(
+//region buildOperator_functions
+private fun buildOperatorNumNumToNum(functionName: String, operator: String = functionName) =
+        buildOperator<Type.Number, Type.Number, Type.Number>(functionName, operator)
+
+private fun buildOperatorNumNumToBool(functionName: String, operator: String = functionName) =
+        buildOperator<Type.Number, Type.Number, Type.Bool>(functionName, operator)
+
+private fun buildOperatorTxtTxtToBool(functionName: String, operator: String = functionName) =
+        buildOperator<Type.Text, Type.Text, Type.Bool>(functionName, operator)
+
+private fun buildOperatorBoolBoolToBool(functionName: String, operator: String = functionName) =
+        buildOperator<Type.Bool, Type.Bool, Type.Bool>(functionName, operator)
+
+//"Prefix" means it will be prefixed in C++, but postfixed in HCL
+private inline fun<reified P, reified R> buildPrefixOperator(functionName: String, operator: String = functionName)
+        where P : Type, R : Type = buildFunction(
         identifier = functionName,
         parameters = listOf(
-                Parameter("leftHand", Type.Number),
-                Parameter("rightHand", Type.Number)
+                Parameter("operand", P::class.objectInstance!!)
         ),
-        returnType = Type.Number,
+        returnType = R::class.objectInstance!!,
+        body = "return $operator operand;",
+        inLine = true
+)
+
+private inline fun<reified V, reified H, reified R> buildOperator(functionName: String, operator: String = functionName)
+        where V : Type, H : Type, R : Type = buildFunction(
+        identifier = functionName,
+        parameters = listOf(
+                Parameter("leftHand", V::class.objectInstance!!),
+                Parameter("rightHand", H::class.objectInstance!!)
+        ),
+        returnType = R::class.objectInstance!!,
         body = "return leftHand $operator rightHand;",
         inLine = true
 )
+//endregion buildOperator_functions
+
+//region builtInFunctions
+private inline fun<reified P: Type> buildToStringFunction() = buildFunction(
```

Comment:
Since this is only relevant for Arduino, perhaps rename this method to make it clearer.